File: 45. Jump Game II.java
class Solution {
    public int jump(int[] nums) {
        int end = 0;
        int maxDistant = 0;
        int n = nums.length;   
        int res = 0;

        for(int i = 0; i < n - 1; ++i){
            if(maxDistant < i + nums[i]){
                maxDistant = i + nums[i];
            }

            if(maxDistant >= n - 1){
                ++res;
                break;
            }

            if(i == end){
                ++res;
                end = maxDistant;
            }
        }

        return res;
    }
}

File: 45. Jump Game II.py
class Solution:
    def jump(self, nums: List[int]) -> int:
        end = 0
        maxDistant = 0
        n = len(nums)
        res = 0

        for i in range(n - 1):
            if(maxDistant < nums[i] + i):
                maxDistant = nums[i] + i
            
            if(maxDistant >= n - 1):
                res += 1
                break
            
            if(i == end):
                res += 1
                end = maxDistant

        return res
    
